### PR TITLE
Raise exception when Variable.reinforce is called

### LINKED ISSUE #3554: reinforce: throw exception (immediately) that points users at what they should use

### PR DESCRIPTION
**Pull request by colesbury**
Fixes #3554

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/autograd/variable.py [hunk 1]
```diff
@@ -206,20 +206,9 @@ def register_hook(self, hook):
...
-            reward(Tensor): Tensor with per-element rewards. It has to match
-                the device location and shape of Variable's data.
-        """
-        if not isinstance(self.grad_fn, StochasticFunction):
-            raise RuntimeError("reinforce() can be only called on outputs "
-                               "of stochastic functions")
-        self.grad_fn._reinforce(reward)
+        raise RuntimeError(
+            "reinforce() was removed. Use torch.distributions instead.\n"
+            "See http://pytorch.org/docs/master/distributions.html")
```
Comment (apaszke) [inline torch/autograd/variable.py]: Can we link to the specific paragraph that shows how to replace these calls?

Comment (soumith) [reply torch/autograd/variable.py]: maybe we can do:

```
See ...
An example of porting from `reinforce()` to using `torch.distributions` is here: https://github.com/pytorch/examples/pull/249/files
```

wdyt adam?

Comment (apaszke) [reply torch/autograd/variable.py]: I'm fine with anything that will point people **directly** to the snippet that lets them fix their code, but this diff is quite large so it might be hard to parse. Let's say "see <distributions docs>, where in the header you can find an example showing how to implement policy gradient".

**Comment by colesbury** [comment]
Let’s just add a section about migration to the docs. I don’t really like
linking to a GitHub PR.

We can also include a snippet in the error message itself.

**Comment by soumith** [comment]
Let's include a snippet in the PR itself.

**Comment by soumith** [comment]
*Lets include a snippet in the error message itself.

**Comment by apaszke** [comment]
Snippet in error message would be great

**Comment by colesbury** [comment]
I'll make a PR with a snippet in the error message

**Review by soumith — APPROVED** [review]
